src/rgw/driver/sfs/sfs_gc.cc

```diff
@@ -38,8 +62,20 @@ int SFSGC::process() {
   lsfs_dout(this, 10) << "garbage collection: processing with max_objects = "
                       << max_objects << dendl;
 
-  // For now, delete only the objects with deleted bucket.
-  process_deleted_buckets();
+  try {
+    process_deleted_buckets();
+    process_deleted_versions();
+  } catch (const sfs::ObjectDeleter::MetadataError& e) {
+    lsfs_dout(this, 30) << e.what() << ". Retrying next iteration" << dendl;
+    return -EIO;
+  } catch (const sfs::ObjectDeleter::FilesystemError& e) {
+    lsfs_dout(this, 30) << e.what() << ". Orphaned files may exists." << dendl;
+    return -EIO;
+  } catch (const BucketMetadataError& e) {
+    lsfs_dout(this, 30) << e.what() << ". Retrying next iteration" << dendl;
+    return -EIO;
+  }
```

Comment:
Can the error handling be further down to where the error occurs?

MetadataError may for example occur if a new version appears while the delete transaction runs between the delete all versions and delete object. Bubbling up that error to here would fail the whole iteration, where only a single object delete would need a retry in the next one

Reply:
It was intentional, to avoid retrying for other objects in this iteration and to keep code as simple as possible.
This code is no longer valid as exceptions have been deleted.